src/domain_extractor.coffee

```diff
@@ -0,0 +1,106 @@
+    Finally, we always drop the www prefix.
+
+    This approach might be problematic in the future, but the only safe
+    alternative is to use the Public Suffix List which contains all
+    the known TLDs. However is quite large in size.
+
+    @see https://github.com/wrangr/psl
+  ###
+  class DomainExtractor
+    constructor: (hostname) ->
+      @hostname = hostname.toLowerCase();
+
+    ###
+      Retrieve the base domain from the given hostname
+
+      @param [Boolean] wildcard Whether to append a dot in the beginning of the domain.
+      @return [String] The base domain
+    ###
+    get: (wildcard = true) ->
+      @domainParts = @hostname.split('.');
+      @domainParts.shift() if @domainParts[0] == 'www' # remove the www prefix
+
+      matched_tld = @_matched_custom_tld()
+      domain = if matched_tld then @_construct_matched_domain(matched_tld) else @_construct_domain()
+      domain = '.' + domain if wildcard
+
+      domain
+
+    ###
+      Construct the domain name from the domain parts
+
+      @return [String] The domain
+    ###
+    _construct_domain: ->
+      domain = [@domainParts.pop()]
+
+      while @domainParts.length
+        domain.unshift(@domainParts.pop())
+        break if domain[0].length > 3
+
+      domain.join('.')
+
+    ###
+      Construct the domain based on the custom tld that was matched in the hostname
+      Keep the custom tld and extract the next domain levels if necessary
+
+      @param [String] matched_tld The custom tld that was matched in the hostname
+      @return [String] The domain
+    ###
+    _construct_matched_domain: (matched_tld) ->
+      matched_tld_length = matched_tld.split('.').length
+      start = @domainParts.length - matched_tld_length
+
+      # if another subdomain level exists, included it the domain
+      if start > 0
+        --start
+        ++matched_tld_length
+
+      @domainParts.splice(start, matched_tld_length).join('.')
+
+    ###
+      Check if the given hostname ends with one of the custom ltd we have set
+
+      @return [String, null] The matched tld or null if not matched
+    ###
+    _matched_custom_tld: ->
+      matched_tld = null
+
+      for tld in @_custom_tlds()
+        dotted_tld = '.' + tld
+        if @hostname.indexOf(dotted_tld, @hostname.length - dotted_tld.length) != -1
+          matched_tld = tld
+          break
+
+      matched_tld
+
+    ###
+      Get the defined custom tlds from the settings. Sort the tlds by
+      descending length in order to match first a tld that might have more
+      levels and has the same base level with another one.
+      e.g. 'aa.skroutz.gr' and 'skroutz.gr'
+      However there is still a problem if two or more tlds are provided with
+      the same length.
+      e.g. 'bb.skroutz.gr' and 'aa.skroutz.gr'
+
+      @return [String, null] The matched tld or null if not matched
+    ###
+    _custom_tlds: ->
+      Settings.custom_tlds.sort (a, b) ->
+        b.length - (a.length)
+
+
+  DomainExtractor
```

Comment:
[Minor] `return DomainExtractor`

(we follow this pattern in all other classes)

Reply:
We broke that pattern in [Validator](https://github.com/skroutz/analytics.js/blob/master/src/validator.coffee#L53). Let's break it again 😄  !